xcm/src/v1/multilocation.rs

```diff
@@ -36,83 +39,385 @@ use parity_scale_codec::{self, Decode, Encode};
+	}
+
+	/// Creates a new `MultiLocation` with no parents and a single `Parachain` interior junction
+	/// specified by `para_id`.
+	///
+	/// The resulting `MultiLocation` can be interpreted as the child-parachain of the current
+	/// consensus system.
+	pub const fn with_parachain_interior(para_id: u32) -> MultiLocation {
+		MultiLocation { parents: 0, interior: Junctions::X1(Junction::Parachain(para_id)) }
+	}
+
+	/// Whether or not the `MultiLocation` has no parents and has a `Null` interior.
+	pub const fn is_here(&self) -> bool {
+		self.parents == 0 && self.interior.len() == 0
+	}
+
+	/// Return a reference to the interior field.
+	pub fn interior(&self) -> &Junctions {
+		&self.interior
+	}
+
+	/// Return a mutable reference to the interior field.
+	pub fn interior_mut(&mut self) -> &mut Junctions {
+		&mut self.interior
+	}
+
+	/// Returns the number of `Parent` junctions at the beginning of `self`.
+	pub const fn parent_count(&self) -> u8 {
+		self.parents
+	}
+
+	/// Returns the number of parents and junctions in `self`.
+	pub const fn len(&self) -> usize {
+		self.parent_count() as usize + self.interior.len()
+	}
+
+	/// Returns the first interior junction, or `None` if the location is empty or contains only
+	/// parents.
+	pub fn first_interior(&self) -> Option<&Junction> {
+		self.interior.first()
+	}
+
+	/// Returns last junction, or `None` if the location is empty or contains only parents.
+	pub fn last(&self) -> Option<&Junction> {
+		self.interior.last()
+	}
+
+	/// Splits off the first interior junction, returning the remaining suffix (first item in tuple)
+	/// and the first element (second item in tuple) or `None` if it was empty.
+	pub fn split_first_interior(self) -> (MultiLocation, Option<Junction>) {
+		let MultiLocation { parents, interior: junctions } = self;
+		let (suffix, first) = junctions.split_first();
+		let multilocation = MultiLocation { parents, interior: suffix };
+		(multilocation, first)
+	}
+
+	/// Splits off the last junction, returning the remaining prefix (first item in tuple) and the last element
+	/// (second item in tuple) or `None` if it was empty or if `self` only contains parents.
+	pub fn split_last(self) -> (MultiLocation, Option<Junction>) {
+		let MultiLocation { parents, interior: junctions } = self;
+		let (prefix, last) = junctions.split_last();
+		let multilocation = MultiLocation { parents, interior: prefix };
+		(multilocation, last)
+	}
+
+	/// Bumps the parent count up by 1. Returns `Err` in case of overflow.
+	pub fn push_parent(&mut self) -> result::Result<(), ()> {
+		if self.len() >= MAX_MULTILOCATION_LENGTH {
+			return Err(())
+		}
+		self.parents = self.parents.saturating_add(1);
+		Ok(())
+	}
+
+	/// Mutates `self`, suffixing its interior junctions with `new`. Returns `Err` in case of overflow.
+	pub fn push_interior(&mut self, new: Junction) -> result::Result<(), ()> {
+		let mut n = Junctions::Null;
+		mem::swap(&mut self.interior, &mut n);
```

Comment:
that's an interesting piece of code, why don't you directly do `self.interior.pushed_with`, but rather swap and then assign to `self.interior? `

Reply:
okay I see why, `push_with` oconsumes self.

Reply:
It was how `push` was implemented in the old version's `MultiLocation`, and I didn't bother changing its details.